cirq/linalg/decompositions.py

```diff
@@ -264,12 +264,115 @@ def so4_to_magic_su2s(
+                 single_qubit_operations_after: Tuple[np.ndarray, np.ndarray]):
+        """Initializes a decomposition for a two-qubit operation U.
+
+        U = g · (a1 ⊗ a0) · exp(i·(x·XX + y·YY + z·ZZ)) · (b1 ⊗ b0)
+
+        Args:
+            global_phase: g from the above equation.
+            single_qubit_operations_before: b0, b1 from the above equation.
+            interaction_coefficients: x, y, z from the above equation.
+            single_qubit_operations_after: a0, a1 from the above equation.
+        """
+        self.global_phase = global_phase
+        self.single_qubit_operations_before = single_qubit_operations_before
+        self.interaction_coefficients = interaction_coefficients
+        self.single_qubit_operations_after = single_qubit_operations_after
+
+    def _eq_tuple(self):
+        def flatten(x):
+            return tuple(tuple(e.flat) for e in x)
+        return (type(KakDecomposition),
+                self.global_phase,
+                tuple(self.interaction_coefficients),
+                flatten(self.single_qubit_operations_before),
+                flatten(self.single_qubit_operations_after))
+
+    def __eq__(self, other):
+        if not isinstance(other, type(self)):
+            return NotImplemented
+        return self._eq_tuple() == other._eq_tuple()
+
+    def __ne__(self, other):
+        return not self == other
+
+    def __hash__(self):
+        return hash(self._eq_tuple())
+
+    def __repr__(self):
+        return (
+            'cirq.KakDecomposition(\n'
+            '    interaction_coefficients={!r},\n'
+            '    single_qubit_operations_before=(\n'
+            '        {},\n'
+            '        {},\n'
+            '    ),\n'
+            '    single_qubit_operations_after=(\n'
+            '        {},\n'
+            '        {},\n'
+            '    ),\n'
+            '    global_phase={!r})'
+        ).format(
+            self.interaction_coefficients,
+            _numpy_array_repr(self.single_qubit_operations_before[0]),
+            _numpy_array_repr(self.single_qubit_operations_before[1]),
+            _numpy_array_repr(self.single_qubit_operations_after[0]),
+            _numpy_array_repr(self.single_qubit_operations_after[1]),
+            self.global_phase,
+        )
+
+    def _unitary_(self):
+        """Returns the decomposition's two-qubit unitary matrix.
+
+        U = g · (a1 ⊗ a0) · exp(i·(x·XX + y·YY + z·ZZ)) · (b1 ⊗ b0)
+        """
+        before = np.kron(*self.single_qubit_operations_before)
+        after = np.kron(*self.single_qubit_operations_after)
+
+        def interaction_matrix(m: np.ndarray, c: float) -> np.ndarray:
+            return map_eigenvalues(np.kron(m, m),
+                                   lambda v: np.exp(1j * v * c))
+
+        x, y, z = self.interaction_coefficients
+        x_mat = np.array([[0, 1], [1, 0]])
+        y_mat = np.array([[0, -1j], [1j, 0]])
+        z_mat = np.array([[1, 0], [0, -1]])
```

Comment:
Should all of this be computed at initialization rather than each time the unitary matrix is needed?
(Not quite sure which would be more efficient, if these will be re-used often)

Reply:
It is rare to ask for the unitary matrix, so it's not worth caching. You almost always go in the opposite direction. This is more of a debugging-and-testing method.